Better Docusaurus Support

frencojobs opened the pull request:
This PR does -

- [x] Allow not only one preset but all presets with supported plugins such as `@docusaurus/content-docs`
- [x] Outdate the requirement of `MDXComponents` override
- [x] Close #41 
- [x] Add docusaurus's default copy-to-clipboard button which is better in a11y and i18n support
- [x] Include default twoslash CSS which is override-able
- [x] Support `title` argument in the code fence as default preset does
- [x] Close #50
- [x] Discard errors, queries, completions when copying to clipboard, also don't discard new lines

orta commented on packages/remark-shiki-twoslash/src/index.ts:
```diff
@@ -35,6 +45,9 @@ function getHTML(code: string, lang: string, metaString: string, highlighters: H
     results = output.join("\n")
+    if (highlighters.length > 1 && wrapFragments) {
+      results = `<div class="shiki-twoslash-fragment">${results}</div>`
+    }
```
I see, nice thinking

frencojobs replied:
yes, I switched that div with a React fragment when rendering back in React

orta commented on packages/remark-shiki-twoslash/src/index.ts:
```diff
@@ -7,11 +7,21 @@ import visit from "unist-util-visit"
 
+export type Settings = UserConfigSettings & {
+  wrapFragments?: true
+}
```
I think this should probably be added in `shiki-twoslash` so that all the config types are in one place ( even though I totally see the value in having it here)

orta commented:
Structurally seems sound to me 👍🏻

frencojobs commented:
Hi @orta, can you review this please. I've added a default CSS based on your CSS but modified it to use the CSS variables from docusaurus. I've added examples in the `examples/docusaurus`, you can see the results there.

I've also made it so that the fence parser will parse the title and pass it through the meta and render with a `div.code-title` tag. For convenience, the `pre` tag will have a class `.with-title` whenever there's a title in the code block. But I've made them optional and to show up if and only if there's the title in the fence so all the tests should just pass OK. It won't affect any existing stuffs.

This makes the user require nothing but a preset and lets them have a much nicer out-of-the-box experience. No overriding or CSS should be required to make it work.

We might have to update the documentations before deploy though, but I'm afraid you'll be better at that. Take your time. Byee.

frencojobs commented:
This is out of scope for this PR but added a fix to #50 since it was on the way.

frencojobs commented:
Found a bug!

orta commented on packages/docusaurus-preset-shiki-twoslash/docusaurus-theme-shiki-twoslash/theme/CodeBlock/styles.css:
```diff
@@ -0,0 +1,290 @@
+  pre.shiki div.highlight::before {
+    transition: none;
+  }
+}
```
This is the only thing which I'd like to see improvements on, it's not blocking though.

Can we automate copying the twoslash CSS [from this file](https://github.com/shikijs/twoslash/blob/main/packages/shiki-twoslash/style.css) (`/packages/shiki-twoslash/style.css`) into the bottom of this file during a deploy? I'll _definitely_ forget to update this CSS as things change.

frencojobs replied:
I don't think it's gonna work though. My commit just a min ago even modify the default CSS soo much it's basically not the same. Things had to be done to make it work with docusaurus's default infima CSS variables and paddings.

orta replied:
OK, then we'll have to think of this as potential tech debt in that we accept that they could get out of sync 👍🏻

orta commented on packages/shiki-twoslash/src/parseCodeFenceInfo.ts:
```diff
@@ -34,7 +38,14 @@ export function parseCodeFenceInfo(lang: string, fullMetaString: string) {
+  return {
+    languageName,
+    meta,
+  }
```
Seems reasonable to me, this will need to be documented somewhere in the shiki-twoslash README too

orta replied:
I'll do this. Let me know when you're happy

frencojobs replied:
I'm finished. Feel free to take on. Was about to go to bed so good luck.

orta commented:
Will give a last read and deploy after a walk
